File: kfhlog/views/log.py
import math
from pyramid.httpexceptions import HTTPForbidden
from pyramid.view import view_config
from sqlalchemy import func

from ..models import Profile, Group, Qso, Mode, Band, Dxcc
from ..models.dbtools.datatypes import RcvdEnum, ModeEnum, ContinentEnum

from .api import json_api_config


@json_api_config(name='get_log')
def _get_log(request, data):
    dbsession = request.dbsession
    qsos = dbsession.query(Qso)
    count = dbsession.query(func.count(Qso.id))
    if 'call_filter' in data:
        qsos = qsos.filter(Qso.call.like(data['call_filter'] + '%'))
        count = count.filter(Qso.call.like(data['call_filter'] + '%'))
    if 'profile' in data:
        profile = data['profile']
        qsos = qsos.filter_by(profile=profile)
        count = count.filter_by(profile=profile)
    if 'group' in data:
        group = data['group']
        qsos = qsos.filter_by(group=group)
        count = count.filter_by(group=group)

    # if 'datetime_off_filter' in data:
    #    qsos = qsos.filter_by(call=data['call_filter'])

    if 'band_filter' in data:
        qsos = qsos.filter_by(band=data['band_filter'])
        count = count.filter_by(band=data['band_filter'])

    if 'mode_filter' in data:
        qsos = qsos.filter_by(mode=data['mode_filter'])
        count = count.filter_by(mode=data['mode_filter'])

    if 'freq_filter' in data:
        qsos = qsos.filter_by(freq=data['freq_filter'])
        count = count.filter_by(freq=data['freq_filter'])

    if 'name_filter' in data:
        qsos = qsos.filter(Qso.name.like(data['name_filter'] + '%'))
        count = count.filter(Qso.name.like(data['name_filter'] + '%'))

    if 'qth_filter' in data:
        qsos = qsos.filter(Qso.qth.like(data['qth_filter'] + '%'))
        count = count.filter(Qso.qth.like(data['qth_filter'] + '%'))

    if 'gridsquare_filter' in data:
        qsos = qsos.filter(Qso.gridsquare.like(data['gridsquare_filter'] + '%'))
        count = count.filter(Qso.gridsquare.like(data['gridsquare_filter'] + '%'))

    if 'dxcc_filter' in data:
        qsos = qsos.filter_by(dxcc=data['dxcc_filter'])
        count = count.filter_by(dxcc=data['dxcc_filter'])

    if 'cont_filter' in data:
        qsos = qsos.filter_by(cont=data['cont_filter'])
        count = count.filter_by(cont=data['cont_filter'])

    if 'ituz_filter' in data:
        qsos = qsos.filter_by(ituz=data['ituz_filter'])
        count = count.filter_by(ituz=data['ituz_filter'])

    if 'cqz_filter' in data:
        qsos = qsos.filter_by(cqz=data['cqz_filter'])
        count = count.filter_by(cqz=data['cqz_filter'])

    if 'iota_filter' in data:
        qsos = qsos.filter(Qso.iota.like(data['iota_filter'] + '%'))
        count = count.filter(Qso.iota.like(data['iota_filter'] + '%'))

    if 'sota_ref_filter' in data:
        qsos = qsos.filter(Qso.sota_ref.like(data['sota_ref_filter'] + '%'))
        count = count.filter(Qso.sota_ref.like(data['sota_ref_filter'] + '%'))

    if 'state_filter' in data:
        qsos = qsos.filter(Qso.state.like(data['state_filter'] + '%'))
        count = count.filter(Qso.state.like(data['state_filter'] + '%'))

    if 'cnty_filter' in data:
        qsos = qsos.filter(Qso.cnty.like(data['cnty_filter'] + '%'))
        count = count.filter(Qso.cnty.like(data['cnty_filter'] + '%'))

    if 'a_index_filter' in data:
        qsos = qsos.filter_by(a_index=data['a_index_filter'])
        count = count.filter_by(a_index=data['a_index_filter'])

    if 'k_index_filter' in data:
        qsos = qsos.filter_by(k_index=data['k_index_filter'])
        count = count.filter_by(k_index=data['k_index_filter'])

    if 'sfi_filter' in data:
        qsos = qsos.filter_by(sfi=data['sfi_filter'])
        count = count.filter_by(sfi=data['sfi_filter'])

    if 'lotw_qsl_rcvd_filter' in data:
        qsos = qsos.filter_by(lotw_qsl_rcvd=data['lotw_qsl_rcvd_filter'])
        count = count.filter_by(lotw_qsl_rcvd=data['lotw_qsl_rcvd_filter'])

    if 'eqsl_qsl_rcvd_filter' in data:
        qsos = qsos.filter_by(eqsl_qsl_rcvd=data['eqsl_qsl_rcvd_filter'])
        count = count.filter_by(eqsl_qsl_rcvd=data['eqsl_qsl_rcvd_filter'])

    if 'qsl_rcvd_filter' in data:
        qsos = qsos.filter_by(qsl_rcvd=data['qsl_rcvd_filter'])
        count = count.filter_by(qsl_rcvd=data['qsl_rcvd_filter'])

    if 'sortby' in data:
        tmp = data['sortby']
        order_map = {
            'profile': Qso.profile,
            '-profile': Qso.profile.desc(),
            'group': Qso.group,
            '-group': Qso.group.desc(),
            'call': Qso.call,
            '-call': Qso.call.desc(),
            'datetime_on': Qso.datetime_on,
            '-datetime_on': Qso.datetime_on.desc(),
            'datetime_off': Qso.datetime_off,
            '-datetime_off': Qso.datetime_off.desc(),
            'band': Qso.band,
            '-band': Qso.band.desc(),
            'mode': Qso.mode,
            '-mode': Qso.mode.desc(),
            'freq': Qso.freq,
            '-freq': Qso.freq.desc(),
            'name': Qso.name,
            '-name': Qso.name.desc(),
            'qth': Qso.qth,
            '-qth': Qso.qth.desc(),
            'gridsquare': Qso.gridsquare,
            '-gridsquare': Qso.gridsquare.desc(),
            'dxcc': Qso.dxcc,
            '-dxcc': Qso.dxcc.desc(),
            'cont': Qso.cont,
            '-cont': Qso.cont.desc(),
            'ituz': Qso.ituz,
            '-ituz': Qso.ituz.desc(),
            'cqz': Qso.cqz,
            '-cqz': Qso.cqz.desc(),
            'iota': Qso.iota,
            '-iota': Qso.iota.desc(),
            'sota_ref': Qso.sota_ref,
            '-sota_ref': Qso.sota_ref.desc(),
            'state': Qso.state,
            '-state': Qso.state.desc(),
            'cnty': Qso.cnty,
            '-cnty': Qso.cnty.desc(),
            'a_index': Qso.a_index,
            '-a_index': Qso.a_index.desc(),
            'k_index': Qso.k_index,
            '-k_index': Qso.k_index.desc(),
            'sfi': Qso.sfi,
            '-sfi': Qso.sfi.desc(),
            'lotw_qsl_rcvd': Qso.lotw_qsl_rcvd,
            '-lotw_qsl_rcvd': Qso.lotw_qsl_rcvd.desc(),
            'eqsl_qsl_rcvd': Qso.eqsl_qsl_rcvd,
            '-eqsl_qsl_rcvd': Qso.eqsl_qsl_rcvd.desc(),
            'qsl_rcvd': Qso.qsl_rcvd,
            '-qsl_rcvd': Qso.qsl_rcvd.desc(),
        }
        if tmp in order_map:
            qsos = qsos.order_by(order_map[tmp])

    count = count.scalar()
    pagesize = 1
    if 'pagesize' in data:
        try:
            pagesize = int(data['pagesize'])
            qsos = qsos.limit(pagesize)
            if 'page' in data:
                page = int(data['page'])
                qsos = qsos.offset((page-1)*pagesize)
        except ValueError as e:
            return {'status': 'error', 'response': 'ValueError'}
    return {'status': 'ok', 'log': qsos.all(), 'count': count, 'pagecount': math.ceil(count/pagesize)}


@json_api_config(name='set_log_preset')
def _set_log_preset(request, data):
    preset = {}
    if 'profile' in data:
        request.session['profile_preset'] = data['profile']
    if 'group' in data:
        request.session['group_preset'] = data['group']

    if 'show_profile' in data:
        preset['show_profile'] = data['show_profile']
    if 'show_group' in data:
        preset['show_group'] = data['show_group']
    if 'show_datetime_off' in data:
        preset['show_datetime_off'] = data['show_datetime_off']
    if 'show_band' in data:
        preset['show_band'] = data['show_band']
    if 'show_mode' in data:
        preset['show_mode'] = data['show_mode']
    if 'show_freq' in data:
        preset['show_freq'] = data['show_freq']
    if 'show_rst_rcvd' in data:
        preset['show_rst_rcvd'] = data['show_rst_rcvd']
    if 'show_rst_send' in data:
        preset['show_rst_send'] = data['show_rst_send']
    if 'show_name' in data:
        preset['show_name'] = data['show_name']
    if 'show_qth' in data:
        preset['show_qth'] = data['show_qth']
    if 'show_gridsquare' in data:
        preset['show_gridsquare'] = data['show_gridsquare']
    if 'show_dxcc' in data:
        preset['show_dxcc'] = data['show_dxcc']
    if 'show_cont' in data:
        preset['show_cont'] = data['show_cont']
    if 'show_ituz' in data:
        preset['show_ituz'] = data['show_ituz']
    if 'show_cqz' in data:
        preset['show_cqz'] = data['show_cqz']
    if 'show_iota' in data:
        preset['show_iota'] = data['show_iota']
    if 'show_sota_ref' in data:
        preset['show_sota_ref'] = data['show_sota_ref']
    if 'show_state' in data:
        preset['show_state'] = data['show_state']
    if 'show_cnty' in data:
        preset['show_cnty'] = data['show_cnty']
    if 'show_a_index' in data:
        preset['show_a_index'] = data['show_a_index']
    if 'show_k_index' in data:
        preset['show_k_index'] = data['show_k_index']
    if 'show_sfi' in data:
        preset['show_sfi'] = data['show_sfi']
    if 'show_lotw_qsl_rcvd' in data:
        preset['show_lotw_qsl_rcvd'] = data['show_lotw_qsl_rcvd']
    if 'show_eqsl_qsl_rcvd' in data:
        preset['show_eqsl_qsl_rcvd'] = data['show_eqsl_qsl_rcvd']
    if 'show_qsl_rcvd' in data:
        preset['show_qsl_rcvd'] = data['show_qsl_rcvd']

    request.session['log_preset'] = preset
    return {'status': 'ok'}


@view_config(route_name='log', renderer='log.jinja2', permission='authenticated')
def log_view(request):
    preset = {
        'show_profile': False,
        'show_group': False,
        'show_datetime_off': False,
        'show_band': True,
        'show_mode': True,
        'show_freq': False,
        'show_rst_rcvd': True,
        'show_rst_send': True,
        'show_name': False,
        'show_qth': False,
        'show_gridsquare': False,
        'show_dxcc': True,
        'show_cont': False,
        'show_ituz': False,
        'show_cqz': False,
        'show_iota': False,
        'show_sota_ref': False,
        'show_state': False,
        'show_cnty': False,
        'show_a_index': False,
        'show_k_index': False,
        'show_sfi': False,
        'show_lotw_qsl_rcvd': False,
        'show_eqsl_qsl_rcvd': False,
        'show_qsl_rcvd': False,
    }
    if 'log_preset' in request.session:
        preset.update(request.session['log_preset'])

    if 'profile' in request.GET:
        preset['profile'] = int(request.GET['profile'])
    elif 'profile_preset' in request.session:
        preset['profile'] = request.session['profile_preset']
    if 'group' in request.GET:
        preset['group'] = int(request.GET['group'])
    elif 'group_preset' in request.session:
        preset['group'] = request.session['group_preset']

    return {'profiles': request.dbsession.query(Profile).all(),
            'groups': request.dbsession.query(Group).all(),
            'mode_all': request.dbsession.query(Mode).all(),
            'band_all': request.dbsession.query(Band).all(),
            'dxcc_all': request.dbsession.query(Dxcc.id, Dxcc.name).order_by(Dxcc.name).all(),
            'cont_all': [a.name for a in ContinentEnum],
            'preset': preset}

File: kfhlog/static/js/log.js
function null_formatter(val) {
    if ( typeof(val) !== "undefined" && val !== null )
        return val;
    return '';
}

function load_log() {
    obj = {}
    /*if($("form#newqso #date_on").val().trim() != "" && $("form#newqso #time_on").val().trim() != "")
        obj["datetime_on"] = $("form#newqso #date_on").val().trim() + 'T' +$("form#newqso #time_on").val().trim();
    if($("form#newqso #comment").val().trim() != "")
        obj["comment"] = $("form#newqso #comment").val().trim();
    */
    $('tr#filters input[type=text]').each(function() {
        if(this.value.trim() != "")
            obj[this.id] = this.value.trim();
    });
    /*$('form#newqso input[type=number]').each(function() {
        parfun = parseInt;
        if (this.id == 'freq' || this.id == 'freq_rx' || this.id == 'tx_pwr')
            parfun = parseFloat;
        if(parfun(this.value.trim()) >= 0)
            obj[this.id] = parfun(this.value.trim());
    });
    $('form#newqso input[type=hidden]').each(function() {
        if(parseInt(this.value.trim()) >= 0)
            obj[this.id] = parseInt(this.value.trim());
    });*/
    $('tr#filters select').each(function() {
        if (this.id == 'cont_filter' && this.value.trim() != '')
            obj[this.id] = this.value.trim();
        else {
            if(parseInt(this.value.trim()) >= 0)
                obj[this.id] = parseInt(this.value.trim());
        }
    });

    obj["sortby"] = $("#sortby").val();

    obj["pagesize"] = $('#pagesize').val();
    obj["page"] = $('#page').val();

    if(parseInt($("#profile").val().trim()) >= 0)
        obj["profile"] = parseInt($("#profile").val().trim());
    if(parseInt($("#group").val().trim()) >= 0)
        obj["group"] = parseInt($("#group").val().trim());

    $.ajax({
	    type:'POST',
		url: '/api/get_log',
		data: JSON.stringify(obj),
		contentType: 'application/json; charset=utf-8',
		success: function(jsn) {
			if( jsn.status != 'ok')
	            return;
	        $("#log > tbody").empty();
            $.each(jsn.log, function(i, item) {
                $('#log > tbody').append('<tr onclick="open_qso(' + item.id+ ')"><td>' +
                item.profile_name + '</td><td>' +
                item.group_name + '</td><td>' +
                item.call + '</td><td>' +
                null_formatter(item.datetime_on).replace('T', ' ').replace('Z', '')  + '</td><td>' +
                null_formatter(item.datetime_off).replace('T', ' ').replace('Z', '') + '</td><td>' +
                item.band_name + '</td><td>' +
                item.mode_name + '</td><td>' +
                item.freq + '</td><td>' +
                item.rst_rcvd + '</td><td>' +
                item.rst_sent + '</td><td>' +
                item.name + '</td><td>' +
                item.qth + '</td><td>' +
                item.gridsquare + '</td><td>' +
                null_formatter(item.dxcc_name) + '</td><td>' +
                null_formatter(item.cont) + '</td><td>' +
                null_formatter(item.ituz) + '</td><td>' +
                null_formatter(item.cqz) + '</td><td>' +
                item.iota + '</td><td>' +
                item.sota_ref + '</td><td>' +
                item.state + '</td><td>' +
                item.cnty + '</td><td>' +
                null_formatter(item.a_index) + '</td><td>' +
                null_formatter(item.k_index) + '</td><td>' +
                null_formatter(item.sfi) + '</td><td>' +
                item.lotw_qsl_rcvd + '</td><td>' +
                item.eqsl_qsl_rcvd + '</td><td>' +
                item.qsl_rcvd + '</td></tr>');
            });
            $('#pagecount').val(jsn.pagecount);
            $('.showcolcheckbox').each(function() {
                var e = jQuery.Event('change');
                $(this).trigger(e, [false]);
            });
		},
		error: function(jqXHR, textStatus, errorThrown) {
      		alert( "Bad request: " + jqXHR.responseText);
    	},
	});
}

function changesort(sort_by) {
    $('tr#headers span').each(function() {
        $(this).removeClass('icon-sort-alpha-asc');
        $(this).removeClass('icon-sort-alpha-desc');
    });

    if ($("#sortby").val() == sort_by) {
        $('tr#headers span#' + sort_by + '_sort').addClass('icon-sort-alpha-desc');
        $("#sortby").val('-' + sort_by);
    }
    else {
        $('tr#headers span#' + sort_by + '_sort').addClass('icon-sort-alpha-asc');
        $("#sortby").val(sort_by);
    }
    $("#sortby").change();
}

function update_preset() {
    preset = {}
    preset['profile'] = parseInt($('#profile').val().trim());
    preset['group'] = parseInt($('#group').val().trim());
    preset['show_profile'] = $('#show_profile').is(':checked');
    preset['show_group'] = $('#show_group').is(':checked');
    preset['show_datetime_off'] = $('#show_datetime_off').is(':checked');
    preset['show_band'] = $('#show_band').is(':checked');
    preset['show_mode'] = $('#show_mode').is(':checked');
    preset['show_freq'] = $('#show_freq').is(':checked');
    preset['show_rst_rcvd'] = $('#show_rst_rcvd').is(':checked');
    preset['show_rst_send'] = $('#show_rst_send').is(':checked');
    preset['show_name'] = $('#show_name').is(':checked');
    preset['show_qth'] = $('#show_qth').is(':checked');
    preset['show_gridsquare'] = $('#show_gridsquare').is(':checked');
    preset['show_dxcc'] = $('#show_dxcc').is(':checked');
    preset['show_cont'] = $('#show_cont').is(':checked');
    preset['show_ituz'] = $('#show_ituz').is(':checked');
    preset['show_cqz'] = $('#show_cqz').is(':checked');
    preset['show_iota'] = $('#show_iota').is(':checked');
    preset['show_sota_ref'] = $('#show_sota_ref').is(':checked');
    preset['show_state'] = $('#show_state').is(':checked');
    preset['show_cnty'] = $('#show_cnty').is(':checked');
    preset['show_a_index'] = $('#show_a_index').is(':checked');
    preset['show_k_index'] = $('#show_k_index').is(':checked');
    preset['show_sfi'] = $('#show_sfi').is(':checked');
    preset['show_lotw_qsl_rcvd'] = $('#show_lotw_qsl_rcvd').is(':checked');
    preset['show_eqsl_qsl_rcvd'] = $('#show_eqsl_qsl_rcvd').is(':checked');
    preset['show_qsl_rcvd'] = $('#show_qsl_rcvd').is(':checked');
    $.ajax({
	    type:'POST',
		url: '/api/set_log_preset',
		data: JSON.stringify(preset),
		contentType: 'application/json; charset=utf-8',
	});
}

function profilegroup_change() {
    load_log();
    update_preset();
}

$(document).ready(function() {
    changesort('datetime_on');
    changesort('datetime_on');
    $('#page').val(1);

    $("#profile").change(profilegroup_change);
    $("#group").change(profilegroup_change);

    $('tr#filters input').each(function() {
        $(this).change(load_log);
    });
    $('tr#filters select').each(function() {
        $(this).change(load_log);
    });

    $("#sortby").change(load_log);

    $('#pagesize').change(load_log);
    $('#page').change(load_log);

    $('#first').click(function() {
        $('#page').val(1);
        $('#page').change();
    });
    $('#prev').click(function() {
        if (parseInt($('#page').val()) > 1) {
            $('#page').val(parseInt($('#page').val())-1);
            $('#page').change();
        }
    });
    $('#next').click(function() {
        $('#page').val(parseInt($('#page').val())+1);
        $('#page').change();
    });
    $('#last').click(function() {
        $('#page').val(parseInt($('#pagecount').val()));
        $('#page').change();
    });

    function make_show_checkbox(selector, col) {
        $(selector).change(function (event, up_preset = true) {
            if ($(selector).is(':checked')) {
                $('#log thead tr td:nth-child(' + col + ')').show();
                $('#log thead tr th:nth-child(' + col + ')').show();
                $('#log tbody tr td:nth-child(' + col + ')').show();
                $('#log tbody tr th:nth-child(' + col + ')').show();
            }
            else {
                $('#log thead tr td:nth-child(' + col + ')').hide();
                $('#log thead tr th:nth-child(' + col + ')').hide();
                $('#log tbody tr td:nth-child(' + col + ')').hide();
                $('#log tbody tr th:nth-child(' + col + ')').hide();
            }
            if(up_preset) update_preset();
        });
        $(selector).attr("checked", $(selector).is(':checked'));
    }

    col=1
    make_show_checkbox("#show_profile", col++);
    make_show_checkbox("#show_group", col++);
    col+=2
    make_show_checkbox("#show_datetime_off", col++);
    make_show_checkbox("#show_band", col++);
    make_show_checkbox("#show_mode", col++);
    make_show_checkbox("#show_freq", col++);
    make_show_checkbox("#show_rst_rcvd", col++);
    make_show_checkbox("#show_rst_send", col++);
    make_show_checkbox("#show_name", col++);
    make_show_checkbox("#show_qth", col++);
    make_show_checkbox("#show_gridsquare", col++);
    make_show_checkbox("#show_dxcc", col++);
    make_show_checkbox("#show_cont", col++);
    make_show_checkbox("#show_ituz", col++);
    make_show_checkbox("#show_cqz", col++);
    make_show_checkbox("#show_iota", col++);
    make_show_checkbox("#show_sota_ref", col++);
    make_show_checkbox("#show_state", col++);
    make_show_checkbox("#show_cnty", col++);
    make_show_checkbox("#show_a_index", col++);
    make_show_checkbox("#show_k_index", col++);
    make_show_checkbox("#show_sfi", col++);
    make_show_checkbox("#show_lotw_qsl_rcvd", col++);
    make_show_checkbox("#show_eqsl_qsl_rcvd", col++);
    make_show_checkbox("#show_qsl_rcvd", col++);
    $('.showcolcheckbox').each(function() {
        var e = jQuery.Event('change');
        $(this).trigger(e, [false]);
    });

    load_log();
});
